src/core_plugins/choropleth/public/choropleth_controller.js

```diff
@@ -0,0 +1,133 @@
+import uiModules from 'ui/modules';
+import _ from 'lodash';
+import AggConfigResult from 'ui/vis/agg_config_result';
+import KibanaMap from 'ui/vis_maps/kibana_map';
+import FilterBarFilterBarClickHandlerProvider from 'ui/filter_bar/filter_bar_click_handler';
+import ChoroplethLayer from './choropleth_layer';
+import colorramps from 'ui/vislib/components/color/colormaps';
+import AggResponsePointSeriesTooltipFormatterProvider from './tooltip_formatter';
+
+const module = uiModules.get('kibana/choropleth', ['kibana']);
+module.controller('KbnChoroplethController', function ($scope, $element, Private, getAppState, tilemapSettings) {
+
+  const filterBarClickHandler = Private(FilterBarFilterBarClickHandlerProvider);
+  const containerNode = $element[0];
+  const tooltipFormatter = Private(AggResponsePointSeriesTooltipFormatterProvider);
+
+  let kibanaMap = null;
+  let choroplethLayer = null;
+
+  async function makeKibanaMap() {
+
+    if (!tilemapSettings.isInitialized()) {
+      await tilemapSettings.loadSettings();
+    }
+
+    const minMaxZoom = tilemapSettings.getMinMaxZoom(false);
+    kibanaMap = new KibanaMap(containerNode, minMaxZoom);
+    const url = tilemapSettings.getUrl();
+    const options = tilemapSettings.getTMSOptions();
+    kibanaMap.setBaseLayer({
+      baseLayerType: 'tms',
+      options: { url, ...options }
+    });
+    kibanaMap.addLegendControl();
+  }
+
+  const kibanaMapReady = makeKibanaMap();
+  $scope.$watch('esResponse', async function (response) {
+    kibanaMapReady.then(() => {
+      const metricsAgg = _.first($scope.vis.aggs.bySchemaName.metric);
+      const termAggId = _.first(_.pluck($scope.vis.aggs.bySchemaName.segment, 'id'));
+      let results;
+      if (!response || !response.aggregations) {
+        results = [];
+      } else {
+        const buckets = response.aggregations[termAggId].buckets;
+        results = buckets.map((bucket) => {
+          return {
+            term: bucket.key,
+            value: getValue(metricsAgg, bucket)
+          };
+        });
+      }
+      const options = $scope.vis.params;
+      if (!options.selectedJoinField) {
+        options.selectedJoinField = options.selectedLayer.fields[0];
+      }
+      updateChoroplethLayer(options.selectedLayer.url);
+      choroplethLayer.setMetrics(results, metricsAgg);
+
+
+      if ($scope.vis.aggs.bySchemaName.segment[0]) {
+        const fieldName = $scope.vis.aggs.bySchemaName.segment[0].params.field.name;
+        choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, fieldName);
+      } else {
+        choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, null);
+      }
+
+      kibanaMap.resize();
```

Comment:
we're having subtle bugs with resizing not getting picked up when adding a visualization to the dashboard (e.g. at page load). we're ensuring the that when we add data, the map will fit to the container.